jni/setting.cpp

```diff
@@ -26,54 +26,74 @@ namespace {
-            ("domain", domain)
-            ("settings", parsed);
+                ("domain", domain)
+                ("settings", parsed);
         return true;
-    } catch ( fostlib::exceptions::exception &e ) {
+    } catch (fostlib::exceptions::exception &e) {
         fostlib::log::info(c_module)
-            ("exception", e.what())
-            ("data", e.data());
+                ("exception", e.what())
+                ("data", e.data());
         return false;
     }
 }
 
 
 extern "C" JNIEXPORT bool JNICALL
 Java_com_felspar_android_Setting_fromString(
-    JNIEnv *env, jobject self, jstring jdomain, jstring jsection, jstring jname, jstring jvalue
+        JNIEnv *env, jobject self, jstring jdomain, jstring jsection, jstring jname, jstring jvalue
 ) {
     try {
         fostlib::string
-            domain = fostlib::jni_cast<fostlib::string>(env, jdomain),
-            section = fostlib::jni_cast<fostlib::string>(env, jsection),
-            name = fostlib::jni_cast<fostlib::string>(env, jname),
-            value = fostlib::jni_cast<fostlib::string>(env, jvalue);
+                domain = fostlib::jni_cast<fostlib::string>(env, jdomain),
+                section = fostlib::jni_cast<fostlib::string>(env, jsection),
+                name = fostlib::jni_cast<fostlib::string>(env, jname),
+                value = fostlib::jni_cast<fostlib::string>(env, jvalue);
 
         g_setting.push_back(
-            std::unique_ptr<fostlib::setting<fostlib::json>>(
-                new fostlib::setting<fostlib::json>(domain, section, name, value)));
+                std::unique_ptr<fostlib::setting<fostlib::json>>(
+                        new fostlib::setting<fostlib::json>(domain, section, name, value)));
         fostlib::log::info(c_module)
-            ("domain", domain)
-            ("section", section)
-            ("name", name)
-            ("value", value);
+                ("domain", domain)
+                ("section", section)
+                ("name", name)
+                ("value", value);
         return true;
-    } catch ( fostlib::exceptions::exception &e ) {
+    } catch (fostlib::exceptions::exception &e) {
         fostlib::log::error(c_module)
-            ("exception", e.what())
-            ("data", e.data());
+                ("exception", e.what())
+                ("data", e.data());
         return false;
     }
 }
 
+extern "C" JNIEXPORT jstring JNICALL
+Java_com_felspar_android_Setting_readString(
+        JNIEnv *env, jobject self, jstring jsection, jstring jname
+) {
+    try {
+        fostlib::string
+                section = fostlib::jni_cast<fostlib::string>(env, jsection),
+                name = fostlib::jni_cast<fostlib::string>(env, jname);
+        if (not fostlib::setting<fostlib::string>::exists(section, name))
+            return env->NewStringUTF("");
+        return env->NewStringUTF(
+                fostlib::setting<fostlib::string>::value(section, name).shrink_to_fit());
```

Comment:
Probably you want to grab the JSON here don't you? Then an empty return string can be an error.

`fostlib::json::unparse(fostlib::setting<fostlib::json>(section, name), false)`

Maybe this should be a new method called `readJSON`?

Your other alternative is to pass in a default value to use in the case that something goes wrong